### PR TITLE
Implementing SA1203 Analyzer. 

### LINKED ISSUE #63: SA1203: ConstantsMustAppearBeforeFields
http://www.stylecop.com/docs/SA1203.html

### PR DESCRIPTION
**Pull request by johnkoerner**
Fixes #63

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs [hunk 1]
```diff
@@ -42,7 +45,50 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+
+                bool thisFieldIsConstant = false;
+                foreach (var modifier in field.Modifiers)
+                {
+                    if (modifier.IsKind(SyntaxKind.ConstKeyword))
+                    {
+                        thisFieldIsConstant = true;
+                        break;
+                    }
+                }
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs]: bool thisFieldIsConstant = field.Modifiers.Any(SyntaxKind.ConstKeyword);

File: StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs [hunk 2]
```diff
@@ -42,7 +45,50 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
         /// <inheritdoc/>
         public override void Initialize(AnalysisContext context)
         {
-            // TODO: Implement analysis
+            context.RegisterSyntaxNodeActionHonorExclusions(this.HandleTypeDeclaration, SyntaxKind.ClassDeclaration);
+            context.RegisterSyntaxNodeActionHonorExclusions(this.HandleTypeDeclaration, SyntaxKind.StructDeclaration);
+        }
+
+        private void HandleTypeDeclaration(SyntaxNodeAnalysisContext context)
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs]: :bulb:  `static`

File: StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs [hunk 3]
```diff
@@ -42,7 +45,50 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+                {
+                    continue;
+                }
+
+                if (!members[i].IsKind(SyntaxKind.FieldDeclaration))
+                {
+                    continue;
+                }
+
+                var field = members[i] as FieldDeclarationSyntax;
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs]: `field` could be null here?

File: StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs [hunk 4]
```diff
@@ -42,7 +45,50 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
         public override void Initialize(AnalysisContext context)
         {
-            // TODO: Implement analysis
+            context.RegisterSyntaxNodeActionHonorExclusions(this.HandleTypeDeclaration, SyntaxKind.ClassDeclaration);
+            context.RegisterSyntaxNodeActionHonorExclusions(this.HandleTypeDeclaration, SyntaxKind.StructDeclaration);
+        }
+
+        private void HandleTypeDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            var typeDeclaration = context.Node as TypeDeclarationSyntax;
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs]: same question about cast

File: StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs (hunk 3 above)
Comment (johnkoerner) [inline StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs]: I am not sure on this.  I will add the null checks just to be safe.

File: StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs (hunk 1 above)
Comment (johnkoerner) [inline StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs]: Nice catch.  Makes that much easier to read.

File: StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs [hunk 5]
```diff
@@ -42,7 +45,53 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+            {
+                return;
+            }
+
+            var members = typeDeclaration.Members;
+            bool nonConstFieldFound = false;
+
+            for (int i = 0; i < members.Count; i++)
+            {
+                if (members[i].IsKind(SyntaxKind.IncompleteMember))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs]: :bulb: This check and the one following it are redundant. The only thing you need is this:

```csharp
var field = members[i] as FieldDeclarationSyntax;
if (field == null)
{
  continue;
}
```

File: StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs [hunk 6]
```diff
@@ -42,7 +45,53 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
-            // TODO: Implement analysis
+            context.RegisterSyntaxNodeActionHonorExclusions(HandleTypeDeclaration, SyntaxKind.ClassDeclaration);
+            context.RegisterSyntaxNodeActionHonorExclusions(HandleTypeDeclaration, SyntaxKind.StructDeclaration);
+        }
+
+        private static void HandleTypeDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            var typeDeclaration = context.Node as TypeDeclarationSyntax;
+
+            if (typeDeclaration == null)
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1203ConstantsMustAppearBeforeFields.cs]: If this true that means bug in Roslyn. We shouldn't handle this. Use

    var typeDeclaration = (TypeDeclarationSyntax)context.Node;

**Pull request merged**